File: ai-service/app/services/explanation_generator.py
from app.models.user_context import UserContext
from app.services.openai_client import openai_client
import structlog
from typing import Dict, Any, List

logger = structlog.get_logger()

class ExplanationGenerator:
    """
    Generates match explanations for recommended categories.
    Provides 2-3 concise reasons why a category matches the user's context.
    """
    
    def __init__(self):
        self.client = openai_client
    
    def _build_context_description(self, context: UserContext) -> str:
        """
        Build a human-readable description of the user context.
        
        Args:
            context: User context
            
        Returns:
            str: Description of the user's nomination
        """
        parts = []
        
        if context.organization_name:
            parts.append(f"Organization: {context.organization_name}")
        
        if context.geography:
            parts.append(f"Location: {context.geography}")
        
        if context.org_type:
            parts.append(f"Type: {context.org_type}")
        
        if context.org_size:
            parts.append(f"Size: {context.org_size}")
        
        if context.nomination_subject:
            parts.append(f"Nominating: {context.nomination_subject}")
        
        if context.description:
            parts.append(f"Achievement: {context.description}")
        
        if context.achievement_focus:
            parts.append(f"Focus areas: {', '.join(context.achievement_focus)}")
        
        if context.tech_orientation:
            parts.append(f"Tech orientation: {context.tech_orientation}")
        
        if context.operating_scope:
            parts.append(f"Operating scope: {context.operating_scope}")
        
        return "\n".join(parts)
    
    async def generate_explanation(
        self,
        context: UserContext,
        category: Dict[str, Any]
    ) -> List[str]:
        """
        Generate match reasons for a single category.
        
        Args:
            context: User context
            category: Category information (name, description, eligibility criteria)
            
        Returns:
            list: 2-3 concise match reasons
        """
        context_desc = self._build_context_description(context)
        
        category_name = category.get("category_name", "Unknown Category")
        category_desc = category.get("description", "")
        program_name = category.get("program_name", "")
        
        # Create prompt for explanation generation
        system_prompt = """You are an expert at explaining why Stevie Awards categories match user nominations.
Generate 2-3 concise, specific reasons why this category is a good match.
Each reason should be one sentence and focus on specific alignment between the nomination and category.
Be encouraging and positive."""
        
        user_prompt = f"""User's Nomination:
{context_desc}

Matched Category:
Name: {category_name}
Program: {program_name}
Description: {category_desc}

Generate 2-3 specific reasons why this category matches the user's nomination.
Return only the reasons, one per line, without numbering or bullet points."""
        
        try:
            response = await self.client.chat_completion(
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt}
                ],
                temperature=0.7,
                max_tokens=200
            )
            
            # Parse reasons from response
            reasons = [
                line.strip().lstrip("-•*").strip()
                for line in response.strip().split("\n")
                if line.strip() and not line.strip().startswith("#")
            ]
            
            # Limit to 2-3 reasons
            reasons = reasons[:3]
            
            # Ensure we have at least 2 reasons
            if len(reasons) < 2:
                reasons.append(f"This category aligns with your {context.nomination_subject} nomination.")
            
            logger.info(
                "explanation_generated",
                category_name=category_name,
                reason_count=len(reasons)
            )
            
            return reasons
            
        except Exception as e:
            logger.error(
                "explanation_generation_error",
                error=str(e),
                category_name=category_name
            )
            
            # Fallback reasons
            return [
                f"This category matches your {context.nomination_subject} nomination.",
                f"Your achievement in {context.achievement_focus[0] if context.achievement_focus else 'this area'} aligns well with this category."
            ]
    
    async def generate_explanations(
        self,
        context: UserContext,
        categories: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """
        Generate match explanations for multiple categories.
        
        Args:
            context: User context
            categories: List of category information
            
        Returns:
            list: List of explanations with category_id and match_reasons
        """
        logger.info(
            "generating_explanations",
            category_count=len(categories)
        )
        
        explanations = []
        
        for category in categories:
            category_id = category.get("category_id")
            
            try:
                match_reasons = await self.generate_explanation(context, category)
                
                explanations.append({
                    "category_id": category_id,
                    "match_reasons": match_reasons
                })
                
            except Exception as e:
                logger.error(
                    "category_explanation_error",
                    error=str(e),
                    category_id=category_id
                )
                # Continue with other categories
                continue
        
        logger.info(
            "explanations_complete",
            success_count=len(explanations)
        )
        
        return explanations

# Global instance
explanation_generator = ExplanationGenerator()

File: api/src/services/explanationGenerator.ts
import { openaiService } from './openaiService';
import logger from '../utils/logger';

/**
 * Explanation Generator Service
 * 
 * Generates match explanations for category recommendations using OpenAI.
 * Replaces the Python AI service explanation generation.
 */

const SYSTEM_PROMPT = `You are an expert at explaining why Stevie Award categories match a nomination.

Your job is to analyze the user's nomination context and explain why each category is a good match.

For each category, provide 2-3 specific, concise reasons why it matches their nomination.

Focus on:
- Alignment with their achievement/description
- Relevance to their organization type and size
- Match with their focus areas
- Fit with the category's purpose

Keep reasons short (1 sentence each) and specific to their nomination.`;

export class ExplanationGenerator {
  /**
   * Generate match explanations for categories
   */
  async generateExplanations(params: {
    userContext: any;
    categories: Array<{
      category_id: string;
      category_name: string;
      description: string;
      program_name: string;
    }>;
  }): Promise<{
    explanations: Array<{
      category_id: string;
      match_reasons: string[];
    }>;
  }> {
    const { userContext, categories } = params;

    logger.info('generating_explanations', {
      category_count: categories.length,
    });

    try {
      const userPrompt = this.buildUserPrompt(userContext, categories);

      const response = await openaiService.chatCompletion({
        messages: [
          { role: 'system', content: SYSTEM_PROMPT },
          { role: 'user', content: userPrompt },
        ],
        temperature: 0.3,
        maxTokens: 2000, // Increased for more categories
      });

      const explanations = this.parseResponse(response, categories);

      logger.info('explanations_generated', {
        count: explanations.length,
      });

      return { explanations };
    } catch (error: any) {
      logger.error('explanation_generation_error', { error: error.message });
      
      // Return empty explanations on error (non-critical)
      return { explanations: [] };
    }
  }

  /**
   * Build user prompt with context and categories
   */
  private buildUserPrompt(
    userContext: any,
    categories: Array<{
      category_id: string;
      category_name: string;
      description: string;
      program_name: string;
    }>
  ): string {
    const parts: string[] = [];

    // User context
    parts.push('User Nomination Context:');
    if (userContext.nomination_subject) {
      parts.push(`- Nominating: ${userContext.nomination_subject}`);
    }
    if (userContext.description) {
      parts.push(`- Achievement: ${userContext.description.substring(0, 200)}`);
    }
    if (userContext.org_type) {
      parts.push(`- Organization Type: ${userContext.org_type}`);
    }
    if (userContext.org_size) {
      parts.push(`- Organization Size: ${userContext.org_size}`);
    }
    if (userContext.achievement_focus && userContext.achievement_focus.length > 0) {
      parts.push(`- Focus Areas: ${userContext.achievement_focus.join(', ')}`);
    }

    parts.push('');
    parts.push('Categories to Explain:');
    parts.push('');

    // Categories
    for (const category of categories) {
      parts.push(`Category ID: ${category.category_id}`);
      parts.push(`Name: ${category.category_name}`);
      parts.push(`Program: ${category.program_name}`);
      parts.push(`Description: ${category.description.substring(0, 150)}`);
      parts.push('');
    }

    parts.push('For each category, provide 2-3 specific reasons why it matches this nomination.');
    parts.push('');
    parts.push('Format your response as JSON:');
    parts.push('{');
    parts.push('  "explanations": [');
    parts.push('    {');
    parts.push('      "category_id": "cat_123",');
    parts.push('      "match_reasons": [');
    parts.push('        "Reason 1",');
    parts.push('        "Reason 2",');
    parts.push('        "Reason 3"');
    parts.push('      ]');
    parts.push('    }');
    parts.push('  ]');
    parts.push('}');

    return parts.join('\n');
  }

  /**
   * Parse OpenAI response with better error handling for truncated JSON
   */
  private parseResponse(
    raw: string,
    categories: Array<{ category_id: string }>
  ): Array<{ category_id: string; match_reasons: string[] }> {
    let text = raw.trim();

    // Strip markdown fences
    if (text.startsWith('```')) {
      const parts = text.split('```');
      text = parts[1] || text;
      if (text.toLowerCase().startsWith('json')) {
        text = text.substring(4);
      }
      text = text.trim();
    }

    try {
      const result = JSON.parse(text);

      if (result.explanations && Array.isArray(result.explanations)) {
        return result.explanations
          .filter((exp: any) => exp.category_id && Array.isArray(exp.match_reasons))
          .map((exp: any) => ({
            category_id: exp.category_id,
            match_reasons: exp.match_reasons.filter((r: any) => typeof r === 'string'),
          }));
      }

      return [];
    } catch (error: any) {
      logger.error('explanation_parse_error', {
        error: error.message,
        response: text.substring(0, 500),
        response_length: text.length,
      });

      // Try to salvage partial JSON by finding complete category objects
      try {
        const partialMatch = text.match(/"explanations"\s*:\s*\[([\s\S]*)/);
        if (partialMatch) {
          // Find all complete category objects
          const categoryPattern = /\{\s*"category_id"\s*:\s*"([^"]+)"\s*,\s*"match_reasons"\s*:\s*\[((?:[^[\]]*|\[[^\]]*\])*)\]\s*\}/g;
          const matches = [...text.matchAll(categoryPattern)];
          
          if (matches.length > 0) {
            logger.info('salvaged_partial_explanations', { count: matches.length });
            return matches.map(match => {
              const categoryId = match[1];
              const reasonsStr = match[2];
              const reasons = reasonsStr
                .split(',')
                .map(r => r.trim().replace(/^"|"$/g, ''))
                .filter(r => r.length > 0);
              return {
                category_id: categoryId,
                match_reasons: reasons,
              };
            });
          }
        }
      } catch (salvageError: any) {
        logger.warn('salvage_failed', { error: salvageError.message });
      }

      // Fallback: generate generic reasons
      logger.info('using_generic_explanations', { count: categories.length });
      return categories.map((cat) => ({
        category_id: cat.category_id,
        match_reasons: [
          'Aligns with your nomination focus',
          'Matches your organization profile',
          'Relevant to your achievement',
        ],
      }));
    }
  }
}

// Export singleton instance
export const explanationGenerator = new ExplanationGenerator();
